File: plugins/harness/skills/harness-install/scripts/install_harness/paths.py
from __future__ import annotations

from pathlib import Path

from .errors import fail
from .models import InstallCandidate


def normalize_requested_target_path(requested_path: str) -> str:

    normalized = (
        requested_path[2:] if requested_path.startswith("./") else requested_path
    )
    reject_unsafe_relative_path(normalized)
    if normalized.endswith("/"):
        fail(f"unsafe target path: {normalized} (must be a file path, not a directory)")
    return normalized


def reject_unsafe_relative_path(path: str) -> None:

    unsafe_path = path[2:] if path.startswith("./") else path
    if unsafe_path in {"", "."} or unsafe_path.startswith("/"):
        fail(
            f"unsafe target path: {path} (must be relative, non-empty, no .. references)"
        )
    parts = unsafe_path.split("/")
    if any(part in {"", ".", ".."} for part in parts):
        fail(
            f"unsafe target path: {path} (must be relative, non-empty, no .. references)"
        )


def ensure_safe_parent_dir(path: str) -> None:

    parent = Path(path).parent
    if str(parent) in {"", "."}:
        return
    reject_unsafe_relative_path(parent.as_posix())
    current = Path()
    for part in parent.parts:
        current = current / part
        if current.is_symlink():
            fail(
                f"[safe_parent] refusing symlink directory component: {current.as_posix()}"
            )
        if current.exists() and not current.is_dir():
            fail(
                f"[safe_parent] parent component is not a directory: {current.as_posix()}"
            )
    parent.mkdir(parents=True, exist_ok=True)


def ensure_safe_file_destination(path: str) -> None:

    clean_path = path[2:] if path.startswith("./") else path
    reject_unsafe_relative_path(clean_path)
    ensure_safe_parent_dir(clean_path)
    target = Path(clean_path)
    if target.is_symlink():
        fail(f"[safe_destination] refusing symlink file destination: {clean_path}")
    if target.is_dir():
        fail(f"[safe_destination] refusing directory file destination: {clean_path}")


def is_common_skip_path(rel: str) -> bool:

    return rel in {
        "AGENTS.md",
        "CLAUDE.md",
        "docs/harness/git-hooks/pre-commit",
        "docs/harness/git-hooks/pre-push",
    }


def is_direct_template_entry(rel: str) -> bool:

    return rel.startswith("docs/harness/templates/") and rel.count("/") == 3


def required_selected_path(config) -> str:

    if config.selected_path is None:
        fail(f"--{config.action} requires a path argument.")
    return config.selected_path


def required_src(candidate: InstallCandidate) -> Path:

    if candidate.src is None:
        fail(f"candidate has no source: {candidate.dst}")
    return candidate.src


def required_real_target(candidate: InstallCandidate) -> str:

    if candidate.real_target is None:
        fail(f"candidate has no root contract target: {candidate.dst}")
    return candidate.real_target

File: plugins/harness/skills/harness-install/scripts/install-harness/paths.ts
import { lstat, mkdir } from "node:fs/promises";
import path from "node:path";

import type { CiHost, InstallCandidate, InstallerConfig } from "./types.js";
import { fail } from "./types.js";

const isEnoent = (error: unknown): boolean =>
  error instanceof Error && "code" in error && error.code === "ENOENT";

const lstatOrAbsent = async (targetPath: string) => {
  try {
    return await lstat(targetPath);
  } catch (error) {
    if (isEnoent(error)) {
      return null;
    }
    throw error;
  }
};

/** Check parent components in order so no lookup crosses a symlink. */
const checkParentComponents = async (
  parents: readonly string[],
  index = 0
): Promise<void> => {
  if (index >= parents.length) {
    return;
  }
  const checkedParent = parents[index];
  if (checkedParent === undefined) {
    return;
  }
  const stat = await lstatOrAbsent(checkedParent);
  if (stat?.isSymbolicLink() === true) {
    fail(
      `[safe_parent] refusing symlink directory component: ${checkedParent}`
    );
  }
  if (stat !== null && !stat.isDirectory()) {
    fail(`[safe_parent] parent component is not a directory: ${checkedParent}`);
  }
  await checkParentComponents(parents, index + 1);
};

export const ensureSafeRelativePath = (targetPath: string): void => {
  const cleanPath = targetPath.startsWith("./")
    ? targetPath.slice(2)
    : targetPath;
  if (
    cleanPath === "" ||
    cleanPath === "." ||
    cleanPath.startsWith("/") ||
    cleanPath.includes("\\")
  ) {
    fail(
      `unsafe target path: ${targetPath} (must be relative, non-empty, no .. references)`
    );
  }
  const parts = cleanPath.split("/");
  if (parts.some((part) => part === "" || part === "." || part === "..")) {
    fail(
      `unsafe target path: ${targetPath} (must be relative, non-empty, no .. references)`
    );
  }
};

export const normalizeRequestedTargetPath = (requestedPath: string): string => {
  const normalized = requestedPath.startsWith("./")
    ? requestedPath.slice(2)
    : requestedPath;
  ensureSafeRelativePath(normalized);
  if (normalized.endsWith("/")) {
    fail(
      `unsafe target path: ${normalized} (must be a file path, not a directory)`
    );
  }
  return normalized;
};

/** Check destination parents without creating them. */
export const checkSafeParentDir = async (targetPath: string): Promise<void> => {
  const parent = path.dirname(targetPath);
  if (parent === "" || parent === ".") {
    return;
  }
  ensureSafeRelativePath(parent);
  let current = "";
  const parents = [];
  for (const part of parent.split("/")) {
    current = current === "" ? part : `${current}/${part}`;
    parents.push(current);
  }
  await checkParentComponents(parents);
};

/** Check a file destination without following or creating path components. */
export const checkSafeFileDestination = async (
  targetPath: string
): Promise<void> => {
  const cleanPath = targetPath.startsWith("./")
    ? targetPath.slice(2)
    : targetPath;
  ensureSafeRelativePath(cleanPath);
  await checkSafeParentDir(cleanPath);
  const stat = await lstatOrAbsent(cleanPath);
  if (stat?.isSymbolicLink() === true) {
    fail(`[safe_destination] refusing symlink file destination: ${cleanPath}`);
  }
  if (stat?.isDirectory()) {
    fail(
      `[safe_destination] refusing directory file destination: ${cleanPath}`
    );
  }
};

/** Check a file destination, then create its missing parent directories. */
export const ensureSafeFileDestination = async (
  targetPath: string
): Promise<void> => {
  await checkSafeFileDestination(targetPath);
  const cleanPath = targetPath.startsWith("./")
    ? targetPath.slice(2)
    : targetPath;
  const parent = path.dirname(cleanPath);
  if (parent !== "" && parent !== ".") {
    await mkdir(parent, { recursive: true });
  }
};

/** Check a parent destination, then create its missing directories. */
export const ensureSafeParentDir = async (
  targetPath: string
): Promise<void> => {
  await checkSafeParentDir(targetPath);
  const parent = path.dirname(targetPath);
  if (parent !== "" && parent !== ".") {
    await mkdir(parent, { recursive: true });
  }
};

export const isCommonSkipPath = (rel: string): boolean =>
  rel === "AGENTS.md" || rel === "CLAUDE.md";

export const isHostTemplatePath = (rel: string, ciHost: CiHost): boolean => {
  if (rel.startsWith(".github/ISSUE_TEMPLATE/")) {
    return ciHost !== "github" && ciHost !== "both";
  }
  if (rel === ".github/pull_request_template.md") {
    return ciHost !== "github" && ciHost !== "both";
  }
  if (rel.startsWith(".gitlab/issue_templates/")) {
    return ciHost !== "gitlab" && ciHost !== "both";
  }
  if (rel.startsWith(".gitlab/merge_request_templates/")) {
    return ciHost !== "gitlab" && ciHost !== "both";
  }
  return false;
};

export const isDirectTemplateEntry = (rel: string): boolean =>
  rel.startsWith("docs/templates/");

export const requiredSelectedPath = (config: InstallerConfig): string => {
  if (config.selectedPath !== null) {
    return config.selectedPath;
  }
  return fail(`--${config.action} requires a path argument.`);
};

export const requiredSrc = (candidate: InstallCandidate): string => {
  if (candidate.src !== undefined) {
    return candidate.src;
  }
  return fail(`candidate has no source: ${candidate.dst}`);
};

export const requiredRealTarget = (candidate: InstallCandidate): string => {
  if (candidate.realTarget !== undefined) {
    return candidate.realTarget;
  }
  return fail(`candidate has no root contract target: ${candidate.dst}`);
};

export const toPosixRelative = (root: string, filePath: string): string =>
  path.relative(root, filePath).split(path.sep).join("/");

export const insideTarget = (root: string, rel: string): string =>
  path.normalize(path.join(root, rel));
